File: raster/src/main/java/com/kjipo/prototype/AngleLine.java
package com.kjipo.prototype;

import com.kjipo.raster.FlowDirection;
import com.kjipo.raster.segment.Pair;
import com.kjipo.raster.segment.Segment;
import com.kjipo.raster.segment.SegmentImpl;
import com.kjipo.segmentation.LineSegmentationKt;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class AngleLine implements AdjustablePrototype {
    private final int id;
    private Pair startPair;
    private double length;
    private double angle;
    private double angleOffset;
    private final Collection<Integer> connectedTo = new HashSet<>();


    private AngleLine(int id, Pair startPair, double length, double angle, Collection<Integer> connectedTo) {
        this.id = id;
        this.startPair = Pair.of(startPair.getRow(), startPair.getColumn());
        this.length = length;
        this.angle = angle;
        this.connectedTo.addAll(connectedTo);
    }

    public AngleLine(int id, Pair startPair, double length, double angle) {
        this(id, startPair, length, angle, Collections.emptySet());
    }

    private AngleLine(int id, Pair startPair, Pair endPair) {
        Objects.requireNonNull(startPair);
        this.id = id;
        int xDelta = endPair.getColumn() - startPair.getColumn();
        int yDelta = endPair.getRow() - startPair.getRow();

        this.startPair = Pair.of(startPair.getRow(), startPair.getColumn());
        this.angle = Math.atan2(yDelta, xDelta);
        this.length = Math.sqrt(Math.pow(xDelta, 2) + Math.pow(yDelta, 2));
    }

    public AngleLine(AngleLine angleLine) {
        this.id = angleLine.id;
        this.startPair = Pair.of(angleLine.startPair.getRow(), angleLine.startPair.getColumn());
        this.length = angleLine.length;
        this.angle = angleLine.angle;
        this.angleOffset = angleLine.angleOffset;
        this.connectedTo.addAll(angleLine.connectedTo);
    }

    @Override
    public Stream<? extends AdjustablePrototype> getMovements() {
        Pair endPair = getEndPair();

        return Arrays.stream(FlowDirection.values())
                .flatMap(flowDirection ->
                        Stream.of(movePair(flowDirection, startPair, endPair, true), movePair(flowDirection, startPair, endPair, false)))
                .filter(angleLine -> angleLine.startPair.getRow() >= 0 && angleLine.startPair.getColumn() >= 0)
                .filter(angleLine -> angleLine.getEndPair().getRow() >= 0 && angleLine.getEndPair().getColumn() >= 0)
                .filter(angleLine -> !angleLine.startPair.equals(angleLine.getEndPair()));
    }

    @Override
    public List<Segment> getSegments() {
        kotlin.Pair newStartPair = new kotlin.Pair<>(startPair.getRow(), startPair.getColumn());
        Pair endPair = getEndPair();
        kotlin.Pair newEndPair = new kotlin.Pair<>(endPair.getRow(), endPair.getColumn());
        List<kotlin.Pair<Integer, Integer>> linePairs = LineSegmentationKt.computeLine(newStartPair, newEndPair);

        return Collections.singletonList(new SegmentImpl(linePairs.stream()
                .map(kotlinPair -> new Pair(kotlinPair.component1(), kotlinPair.component2()))
                .collect(Collectors.toList())));
    }

    public int getId() {
        return id;
    }


    public Pair getEndPair() {
        double xDelta = length * Math.cos(angle + angleOffset);
        double yDelta = length * Math.sin(angle + angleOffset);

        return Pair.of((int) Math.round(startPair.getRow() + yDelta), (int) Math.round(startPair.getColumn() + xDelta));
    }

    public Pair getStartPair() {
        return Pair.of(startPair.getRow(), startPair.getColumn());
    }

    public void setStartPair(Pair startPair) {
        Objects.requireNonNull(startPair);
        this.startPair = Pair.of(startPair.getRow(), startPair.getColumn());
    }

    public double getAngle() {
        return angle;
    }

    public double getLength() {
        return length;
    }

    public void setLength(double length) {
        this.length = length;
    }

    public double getAngleOffset() {
        return angleOffset;
    }

    public void setAngleOffset(double angleOffset) {
        this.angleOffset = angleOffset;
    }

    public void addAngleOffset(double angleOffset) {
        this.angleOffset += angleOffset;
    }

    public Collection<Integer> getConnectedTo() {
        return new HashSet<>(connectedTo);
    }

    public void addConnectedTo(int id) {
        connectedTo.add(id);
    }

    private AngleLine movePair(FlowDirection flowDirection, Pair startPair, Pair endPair, boolean moveStartPair) {
        if (moveStartPair) {
            return new AngleLine(
                    id,
                    Pair.of(startPair.getRow() + flowDirection.getRowShift(),
                            startPair.getColumn() + flowDirection.getColumnShift()),
                    endPair);
        }
        return new AngleLine(id, startPair,
                Pair.of(endPair.getRow() + flowDirection.getRowShift(),
                        endPair.getColumn() + flowDirection.getColumnShift()));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AngleLine angleLine = (AngleLine) o;
        return id == angleLine.id &&
                Double.compare(angleLine.length, length) == 0 &&
                Double.compare(angleLine.angle, angle) == 0 &&
                Double.compare(angleLine.angleOffset, angleOffset) == 0 &&
                Objects.equals(startPair, angleLine.startPair) &&
                Objects.equals(connectedTo, angleLine.connectedTo);
    }

    @Override
    public int hashCode() {

        return Objects.hash(id, startPair, length, angle, angleOffset, connectedTo);
    }

    @Override
    public String toString() {
        return "AngleLine{" +
                "id=" + id +
                ", startPair=" + startPair +
                ", length=" + length +
                ", angle=" + angle +
                ", angleOffset=" + angleOffset +
                ", connectedTo=" + connectedTo +
                ", endPair=" +getEndPair() +
                '}';
    }
}

File: common/src/main/kotlin/representation/line/AngleLine.kt
package com.kjipo.representation.line

import com.kjipo.representation.raster.FlowDirection
import kotlin.math.atan2
import kotlin.math.sqrt


//class AngleLine {
//    val id: Int
//    private var startPair: Pair<Int, Int>
//    var length: Double
//    var angle: Double
//        private set
//    var angleOffset = 0.0
//    private val connectedTo: MutableCollection<Int> = HashSet()
//
//    private constructor(id: Int, startPair:  Pair<Int, Int>, length: Double, angle: Double, connectedTo: Collection<Int>) {
//        this.id = id
//        this.startPair = Pair(startPair.first, startPair.second)
//        this.length = length
//        this.angle = angle
//        this.connectedTo.addAll(connectedTo)
//    }
//
//    constructor(id: Int, startPair:  Pair<Int, Int>, length: Double, angle: Double) : this(id, startPair, length, angle, emptySet<Int>()) {}
//
//    constructor(id: Int, startPair:  Pair<Int, Int>, endPair:  Pair<Int, Int>) {
//        this.id = id
//        val xDelta: Int = endPair.second - startPair.second
//        val yDelta: Int = endPair.first- startPair.first
//        this.startPair =  Pair<Int, Int>.of(startPair.first, startPair.second)
//        angle = atan2(yDelta.toDouble(), xDelta.toDouble())
//        length = sqrt(pow(xDelta.toDouble(), 2.0) + java.lang.Math.pow(yDelta.toDouble(), 2.0))
//    }
//
//    constructor(angleLine: AngleLine) {
//        id = angleLine.id
//        startPair =  Pair(angleLine.startPair.first, angleLine.startPair.second)
//        length = angleLine.length
//        angle = angleLine.angle
//        angleOffset = angleLine.angleOffset
//        connectedTo.addAll(angleLine.connectedTo)
//    }
//
//    val segments: List<Any>
//        get() {
//            val newStartPair: Pair<*, *> = Pair<Int, Int>(startPair.first, startPair.second)
//            val endPair:  Pair<Int, Int> = endPair
//            val newEndPair: Pair<*, *> = Pair<Int, Int>(endPair.first, endPair.second)
//            val linePairs: List<Pair<Int, Int>> = computeLine(newStartPair, newEndPair)
//            return listOf(SegmentImpl(linePairs.stream()
//                    .map(java.util.function.Function<Pair<Int, Int>,  Pair<Int, Int>> { kotlinPair: Pair<Int?, Int?> ->  Pair<Int, Int>(kotlinPair.component1(), kotlinPair.component2()) })
//                    .collect(Collectors.toList())))
//        }
//
//    val endPair:  Pair<Int, Int>
//        get() {
//            val xDelta: Double = length * java.lang.Math.cos(angle + angleOffset)
//            val yDelta: Double = length * java.lang.Math.sin(angle + angleOffset)
//            return  Pair<Int, Int>.of(java.lang.Math.round(startPair.first + yDelta) as Int, java.lang.Math.round(startPair.second + xDelta) as Int)
//        }
//
//    fun getStartPair():  Pair<Int, Int> {
//        return  Pair(startPair.first, startPair.second)
//    }
//
//    fun setStartPair(startPair:  Pair<Int, Int>) {
//        java.util.Objects.requireNonNull(startPair)
//        this.startPair =  Pair<Int, Int>.of(startPair.first, startPair.second)
//    }
//
//    fun addAngleOffset(angleOffset: Double) {
//        this.angleOffset += angleOffset
//    }
//
//    fun getConnectedTo(): Collection<Int> {
//        return HashSet(connectedTo)
//    }
//
//    fun addConnectedTo(id: Int) {
//        connectedTo.add(id)
//    }
//
//    private fun movePair(flowDirection: FlowDirection, startPair:  Pair<Int, Int>, endPair:  Pair<Int, Int>, moveStartPair: Boolean): com.kjipo.prototype.AngleLine {
//        return if (moveStartPair) {
//            com.kjipo.prototype.AngleLine(
//                    id,
//                     Pair<Int, Int>.of(startPair.first + flowDirection.getRowShift(),
//                            startPair.second + flowDirection.getColumnShift()),
//                    endPair)
//        } else com.kjipo.prototype.AngleLine(id, startPair,
//                 Pair<Int, Int>.of(endPair.first + flowDirection.getRowShift(),
//                        endPair.second + flowDirection.getColumnShift()))
//    }
//
//    override fun equals(o: Any?): Boolean {
//        if (this === o) return true
//        if (o == null || javaClass != o.javaClass) return false
//        val angleLine: com.kjipo.prototype.AngleLine = o as com.kjipo.prototype.AngleLine
//        return id == angleLine.id && java.lang.Double.compare(angleLine.length, length) == 0 && java.lang.Double.compare(angleLine.angle, angle) == 0 && java.lang.Double.compare(angleLine.angleOffset, angleOffset) == 0 &&
//                java.util.Objects.equals(startPair, angleLine.startPair) &&
//                java.util.Objects.equals(connectedTo, angleLine.connectedTo)
//    }
//
//    override fun hashCode(): Int {
//        return java.util.Objects.hash(id, startPair, length, angle, angleOffset, connectedTo)
//    }
//
//    override fun toString(): String {
//        return "AngleLine{" +
//                "id=" + id +
//                ", startPair=" + startPair +
//                ", length=" + length +
//                ", angle=" + angle +
//                ", angleOffset=" + angleOffset +
//                ", connectedTo=" + connectedTo +
//                ", endPair=" + endPair +
//                '}'
//    }
//}